src/common/lib/transport/connectionmanager.ts

```diff
@@ -1508,129 +1261,160 @@ class ConnectionManager extends EventEmitter {
   }
 
   /**
-   * There are three stages in connecting:
-   * - preference: if there is a cached transport preference, we try to connect
-   *   on that. If that fails or times out we abort the attempt, remove the
-   *   preference and fall back to base. If it succeeds, we try upgrading it if
-   *   needed (will only be in the case where the preference is xhrs and the
-   *   browser supports ws).
-   * - base: we try to connect with the best transport that we think will
-   *   never fail for this platform. If it doesn't work, we try fallback hosts.
-   * - upgrade: given a connected transport, we see if there are any better
-   *   ones, and if so, try to upgrade to them.
-   *
-   * connectImpl works out what stage you're at (which is purely a function of
-   * the current connection state and whether there are any stored preferences),
-   * and dispatches accordingly. After a transport has been set pending,
-   * tryATransport calls connectImpl to see if there's another stage to be done.
+   * connectImpl works out which transport to initially attempt a connection with
```

Comment:
The previous version documented how the connection sequence worked here, can we not do the same? it's not necessarily obvious

Reply:
added some detailed docstrings to `connectImpl` and `connectWs`